subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintWriter.kt

```diff
@@ -258,18 +262,19 @@ class ConfigurationCacheFingerprintWriter(
     }
 
     override fun envVariablesPrefixedBy(prefix: String, snapshot: Map<String, String?>) {
-        if (inputTrackingDisabledForThread) {
+        if (isInputTrackingDisabled()) {
             return
         }
         buildScopedSink.write(ConfigurationCacheFingerprint.EnvironmentVariablesPrefixedBy(prefix, snapshot))
     }
 
     override fun beforeValueObtained() {
-        inputTrackingDisabledForThread = true
+        // Do not track additional inputs while computing a value of the value source.
+        inputTrackingState.disableForCurrentThread()
     }
 
     override fun afterValueObtained() {
-        inputTrackingDisabledForThread = false
+        inputTrackingState.enableForCurrentThread()
```

Comment:
❓ Won't this cause input tracking to be re-enabled unconditionally after a value is obtained even when input tracking was already disabled before the value was obtained? I wonder if such scenario even exists 🤔

Reply:
It exists: #22305. I was pushing this to the overall fix of obtaining a value source inside the untracked block, but this can be solved separately. I filed #22494 to track the rest. It seems to be too involved for 7.6 line, as it includes serialization changes we discussed last week.